src/tracked.jl

```diff
@@ -270,10 +270,22 @@ Base.convert(::Type{T}, t::T) where {T<:TrackedReal} = t
 Base.convert(::Type{T}, t::T) where {T<:TrackedArray} = t
 
 for R in REAL_TYPES
-    @eval Base.promote_rule(::Type{$R}, ::Type{TrackedReal{V,D,O}}) where {V,D,O} = TrackedReal{promote_type($R,V),D,O}
+    R === :Dual && continue # ForwardDiff.Dual is handled below 
+    @eval begin
+        if isconcretetype($R) # issue ForwardDiff#322
+            Base.promote_rule(::Type{TrackedReal{V,D,O}}, ::Type{$R}) where {V,D,O} = TrackedReal{promote_type(V,$R),D,O}
```

Comment:
Since we are implementing both methods, do we know if they are all covered by tests? Might be worth adding a test checking that switching arguments gives the same result.

Reply:
Both are tested since `promote_type` calls both `promote_rule` definitions.

Reply:
I should add that actually tests fail when one of the definitions is removed due to eg stackoverflow errors when calling `promote_type` as mentioned in the OP.